pallets/capacity/src/lib.rs

```diff
@@ -644,18 +949,292 @@ impl<T: Config> Pallet<T> {
+		} else {
+			boost_history.subtract_era_balance(&current_era, &boost_amount)
+		};
+		match upsert_result {
+			Some(0usize) => ProviderBoostHistories::<T>::remove(account),
+			None => return Err(DispatchError::from(Error::<T>::EraOutOfRange)),
+			_ => ProviderBoostHistories::<T>::set(account, Some(boost_history)),
+		}
+		Ok(())
+	}
+
+	pub(crate) fn has_unclaimed_rewards(account: &T::AccountId) -> bool {
+		let current_era = Self::get_current_era().era_index;
+		match Self::get_staking_history_for(account) {
+			Some(provider_boost_history) => {
+				match provider_boost_history.count() {
+					0usize => false,
+					1usize => {
+						// if there is just one era entry and:
+						// it's for the previous era, it means we've already paid out rewards for that era, or they just staked in the last era.
+						// or if it's for the current era, they only just started staking.
+						provider_boost_history
+							.get_entry_for_era(&current_era.saturating_sub(1u32.into()))
+							.is_none() && provider_boost_history
+							.get_entry_for_era(&current_era)
+							.is_none()
+					},
+					_ => true,
+				}
+			},
+			None => false,
+		} // 1r
+	}
+
+	/// Get all unclaimed rewards information for each eligible Reward Era.
+	/// If no unclaimed rewards, returns empty list.
+	pub fn list_unclaimed_rewards(
+		account: &T::AccountId,
+	) -> Result<
+		BoundedVec<
+			UnclaimedRewardInfo<BalanceOf<T>, BlockNumberFor<T>>,
+			T::ProviderBoostHistoryLimit,
+		>,
+		DispatchError,
+	> {
+		if !Self::has_unclaimed_rewards(account) {
+			return Ok(BoundedVec::new());
 		}
+
+		let staking_history =
+			Self::get_staking_history_for(account).ok_or(Error::<T>::NotAProviderBoostAccount)?; // cached read
+
+		let current_era_info = Self::get_current_era(); // cached read, ditto
+		let max_history: u32 = T::ProviderBoostHistoryLimit::get();
+
+		let mut reward_era = current_era_info.era_index.saturating_sub((max_history).into());
+		let end_era = current_era_info.era_index.saturating_sub(One::one());
+
+		// start with how much was staked in the era before the earliest for which there are eligible rewards.
+		let mut previous_amount: BalanceOf<T> =
+			staking_history.get_amount_staked_for_era(&(reward_era.saturating_sub(1u32.into())));
+
+		let mut unclaimed_rewards: BoundedVec<
+			UnclaimedRewardInfo<BalanceOf<T>, BlockNumberFor<T>>,
+			T::ProviderBoostHistoryLimit,
+		> = BoundedVec::new();
+		while reward_era.le(&end_era) {
+			let staked_amount = staking_history.get_amount_staked_for_era(&reward_era);
+			if !staked_amount.is_zero() {
+				let expires_at_era = reward_era.saturating_add(max_history.into());
+				let expires_at_block = Self::block_at_end_of_era(expires_at_era);
+				let eligible_amount = if staked_amount.lt(&previous_amount) {
+					staked_amount
+				} else {
+					previous_amount
+				};
+				let total_for_era =
+					Self::get_total_stake_for_past_era(reward_era, current_era_info.era_index)?;
```

Comment:
This can get optimized. Instead of reading the storage per era you should read the storage per chunk.